src/urllib3/response.py

```diff
@@ -910,21 +910,17 @@ def read(
         if amt is not None:
             cache_content = False
 
+            if amt == 0:
+                return b""
+
             if len(self._decoded_buffer) >= amt:
                 return self._decoded_buffer.get(amt)
 
-        data = self._raw_read(amt)
-
-        flush_decoder = amt is None or (amt != 0 and not data)
```

Comment:
What do you think about keeping the current code structure for now, but defining `flush_decoder` directly in the `self._decode` calls? I.e., calling `self._decode(data, decode_content, flush_decoder=True)` when `amt is None`  and `self._decode(data, decode_content, flush_decoder=not data)` otherwise.

As per https://github.com/urllib3/urllib3/issues/2799#issuecomment-1901102673, this can be a user-facing change, so it has to be accompanied by new tests and a news entry.

Reply:
There should(!) be no functional change, but naturally I may have overseen something 

I'll try to recreate the squashed commits from a backup (thought this https://github.com/urllib3/urllib3/pull/3262#issuecomment-1884214847 was final, so i squashed), and answer you then.

Reply:
here we go: https://github.com/urllib3/urllib3/pull/3262#discussion_r1460238172

Reply:
> What do you think about keeping the current code structure for now
@illia-v , it is fixed with step https://github.com/urllib3/urllib3/pull/3262#discussion_r1460238172, all other steps should be functionally neutral or minor.

I'm fine with however you proceed, i can try to send a separate PR in, but if you prefer, you can author a PR (and if you worry about crediting, just add something like "based on digging by abebeos within https://github.com/urllib3/urllib3/issues/2799#issuecomment-1872448595" to the commit comment).

You (lead maintainers) must(!) however realize that the cause of all this #2799 (and other issues, incl. security related) is(!) the current code-structure.

Reply:
> What do you think about keeping the current code structure for now

yes, see https://github.com/urllib3/urllib3/pull/3262#issuecomment-1901960676

Reply:
applied the minimum steps

- => https://github.com/urllib3/urllib3/pull/3292